### PR TITLE
sniproxy: Added SNI Proxy package

### PR DESCRIPTION
**Pull request by G-M0N3Y-2503**
Signed-off-by: Gerard Ryan <G.M0N3Y.2503@gmail.com>

Maintainer: @G-M0N3Y-2503
Compile tested: x86_x64, Hyper-V, OpenWrt Master
Run tested: x86_x64, Hyper-V, OpenWrt Master

Description:
Added SNI Proxy package for forthcoming https://github.com/lancachenet/monolithic docker image package

### CODE REVIEW

File: net/sniproxy/files/sniproxy.init [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+	config_foreach cfg_add_resolvers "resolver" "${RUNTIME_CONF_FILE}"
+	config_foreach cfg_add_default_logs "error_log" "${RUNTIME_CONF_FILE}"
+	config_foreach cfg_add_default_logs "access_log" "${RUNTIME_CONF_FILE}"
+	config_foreach cfg_add_listeners "listener" "${RUNTIME_CONF_FILE}"
+	config_foreach cfg_add_tables "table" "${RUNTIME_CONF_FILE}"
+}
+
+start_service() {
+	local CONF_FILE="/etc/config/${SERVICE_NAME}"
+	local RUNTIME_CONF_FILE="/etc/${SERVICE_NAME}.conf"
```
Comment (feckert) [inline net/sniproxy/files/sniproxy.init]: I would suggest that we but the runtime config into a tmp folder to decrease the writes to the NAND.

Comment (G-M0N3Y-2503) [reply net/sniproxy/files/sniproxy.init]: Just thought of this when I woke up, good catch!
Fixed now.

**Comment by nmav** [comment]
What's the reason to bring sniproxy instead of using haproxy?

**Comment by G-M0N3Y-2503** [comment]
> What's the reason to bring sniproxy instead of using haproxy?


The primary reason I didn't use haproxy is that it didn't come up in my research for a Server Name Indication Proxy. That being said, I just tried to configure haproxy as a replacement to sniproxy for [my lancachenet package](https://github.com/G-M0N3Y-2503/packages/tree/feature_lancache) and it only sort of worked, it seems to interfere with steam downloads and makes it cut out frequently.
so unless it's a misconfiguration it doesn't look like it will do the same job.

<details>
<summary>haproxy.conf</summary>

```conf
global
	maxconn 32000
	ulimit-n 65535
	uid 0
	gid 0
	daemon
	nosplice
	nbproc 4

defaults
	timeout connect 5000ms
	mode tcp

resolvers upstream
	nameserver cloudflare-primary 1.1.1.1:53
	nameserver cloudflare-secondary 1.0.0.1:53

listen http_proxy
	bind :80
	use-server LuCI if { hdr(host) -i openwrt.lan }
	server LuCI localhost:8080
	server cache localhost:8081

listen https_proxy
	bind :443
	use-server LuCI if { hdr(host) -i openwrt.lan }
	server LuCI localhost:8443
	server any * resolvers upstream
```
</details>
<details>
<summary>sniproxy.conf</summary>

```conf
config sniproxy
	option username		'nobody'

config error_log
	option syslog		'daemon'
	option priority		'warning'

config access_log
	option syslog		'daemon'
	option priority		'warning'

config listener
	option address		'80'
	option protocol		'http'
	option table		'http'

config listener
	option address		'443'
	option protocol		'tls'
	option table		'https'

config table 'http'
	list entry		'openwrt.lan *:8080'
	list entry		'.* *:8081'

config table 'https'
	list entry		'openwrt.lan *:8443'
	list entry		'.* *:443'
```
</details>

**Comment by nmav** [comment]
> The primary reason I didn't use haproxy is that it didn't come up in my research for a Server Name Indication Proxy. That being said, I just tried to configure haproxy as a replacement to sniproxy for my lancachenet package and it only sort of worked, it seems to interfere with steam downloads and makes it cut out frequently. so unless it's a misconfiguration it doesn't look like it will do the same job.

I'd be you have a misconfiguration. I'm a long time maintainer of sniproxy for fedora, and we gave up on that package in favor of haproxy because of its superior performance and flexibility.

**Comment by G-M0N3Y-2503** [comment]
> > The primary reason I didn't use haproxy is that it didn't come up in my research for a Server Name Indication Proxy. That being said, I just tried to configure haproxy as a replacement to sniproxy for my lancachenet package and it only sort of worked, it seems to interfere with steam downloads and makes it cut out frequently. so unless it's a misconfiguration it doesn't look like it will do the same job.
> 
> I'd bet you have a misconfiguration. I'm a long time maintainer of sniproxy for fedora, and we gave up on that package in favor of haproxy because of its superior performance and flexibility. I see no reason for that package in openwrt other than we can bring it.

It could be argued that sniproxy is easier to configure and much easier to interface with the UCI and more likely to get a LuCI interface. That being said, that's not something I'd like to undertake, so I'll leave this PR open until I can confirm that haproxy is a suitable alternative.

**Comment by G-M0N3Y-2503** [comment]
seems I was just missing `v4v6`, haproxy seems to work so I'll close this PR

**Pull request closed without merging**